lib/AST/TypeCheckRequests.cpp

```diff
@@ -334,6 +334,24 @@ bool RequirementRequest::visitRequirements(
     // Resolve to a requirement.
     auto req = evaluator(RequirementRequest{owner, index, stage});
     if (req) {
+      
+      // If we're extending a protocol and adding a redundant requirement,
+      // for example, `extension Foo where Self: Foo`, then emit a
+      // diagnostic.
+      
+      // FIXME: Leads to duplicate diagnostic emissions
+      if (auto extDecl = dyn_cast<ExtensionDecl>(owner.dc->getAsDecl())) {
```

Comment:
Hm, I would rather not put this check inside visitRequirements, because otherwise like you said, it will emit multiple times.

Maybe you can put this before calling `GenericSignatureBuilder::addRequirement` in `TypeCheckGeneric.cpp`?